ledger/src/blockstore_processor.rs

```diff
@@ -4131,4 +4163,100 @@ pub mod tests {
+            .feature_set
+            .is_active(&feature_set::cap_accounts_data_size_per_block::id()));
+
+        for _ in 0..NUM_ACCOUNTS {
+            let transaction = system_transaction::create_account(
+                &mint_keypair,
+                &Keypair::new(),
+                bank.last_blockhash(),
+                LAMPORTS_PER_SOL,
+                ACCOUNT_SIZE,
+                &solana_sdk::system_program::id(),
+            );
+            let entry = next_entry(&bank.last_blockhash(), 1, vec![transaction]);
+            assert!(process_entries_for_tests(&bank, vec![entry], true, None, None).is_ok());
+        }
+
+        let transaction = system_transaction::create_account(
+            &mint_keypair,
+            &Keypair::new(),
+            bank.last_blockhash(),
+            LAMPORTS_PER_SOL,
+            ACCOUNT_SIZE,
+            &solana_sdk::system_program::id(),
+        );
+        let entry = next_entry(&bank.last_blockhash(), 1, vec![transaction]);
+        assert_eq!(
+            process_entries_for_tests(&bank, vec![entry], true, None, None),
+            Err(TransactionError::WouldExceedAccountDataBlockLimit)
+        );
+    }
+
+    #[test]
+    fn test_check_accounts_data_total_size() {
+        const REMAINING_ACCOUNTS_DATA_SIZE: u64 =
+            MAX_ACCOUNT_DATA_BLOCK_LEN - MAX_PERMITTED_DATA_LENGTH;
+        const INITIAL_ACCOUNTS_DATA_SIZE: u64 =
+            MAX_ACCOUNTS_DATA_LEN - REMAINING_ACCOUNTS_DATA_SIZE;
+        const NUM_ACCOUNTS: u64 = 11;
+        const ACCOUNT_SIZE: u64 = REMAINING_ACCOUNTS_DATA_SIZE / NUM_ACCOUNTS;
+
+        let GenesisConfigInfo {
+            genesis_config,
+            mint_keypair,
+            ..
+        } = create_genesis_config((1_000_000 + NUM_ACCOUNTS + 1) * LAMPORTS_PER_SOL);
+        let mut bank = Bank::new_for_tests(&genesis_config);
+        bank.set_accounts_data_size_initial_for_tests(INITIAL_ACCOUNTS_DATA_SIZE);
+        let bank = Arc::new(bank);
+        assert!(bank
+            .feature_set
+            .is_active(&feature_set::cap_accounts_data_len::id()));
+
+        for _ in 0..NUM_ACCOUNTS {
+            let transaction = system_transaction::create_account(
+                &mint_keypair,
+                &Keypair::new(),
+                bank.last_blockhash(),
+                LAMPORTS_PER_SOL,
+                ACCOUNT_SIZE,
+                &solana_sdk::system_program::id(),
+            );
+            let entry = next_entry(&bank.last_blockhash(), 1, vec![transaction]);
+            assert!(process_entries_for_tests(&bank, vec![entry], true, None, None).is_ok());
+        }
+
+        let transaction = system_transaction::create_account(
+            &mint_keypair,
+            &Keypair::new(),
+            bank.last_blockhash(),
+            LAMPORTS_PER_SOL,
+            ACCOUNT_SIZE,
+            &solana_sdk::system_program::id(),
+        );
+        let entry = next_entry(&bank.last_blockhash(), 1, vec![transaction]);
```

Comment:
Is there a test which has a decrease in account size? How about a resize?

Reply:
Not yet. Good call, I'll do that.

Reply:
@sakridge How important do you feel testing the accounts size _decreasing_ is here?

I've been iterating on ways to test this out and haven't had luck yet. Here are my eight attempts so far:
https://gist.github.com/brooksprumo/5ed10ca0b54d92397be82a4316037e1f

Some things that make decreasing the accounts size challenging:
* First, I planned to create an account with large data size, then transfer its lamports out, thus causing the account to be cleaned up and its data removed. Unfortunately accounts with non-zero data size cannot sign transactions, so this test method is not possible.
* Second, I wanted to call realloc on an account. Realloc is only supported on programs, so calling realloc on a simple account is not possible.
* Third, I wanted to create-and-deploy a program then call the close instruction on it, which should remove its ProgramData account. So far I haven't had luck successfully creating the program; deploying the program goes through a lot of checks, and I'm going so far as to copy the bytes from on-chain `.so`s used in other tests and put them into an array to then copy into my program buffer. This is failing the ELF parsing (currently trying to figure out why; I used `xxd` to get the bytes, so maybe there's some flag I need to use).

At this point I decided to take a step back and reevaluate what I'm really trying to test.

As a simpler option, I could manually change the `accounts_data_size` fields in the test banks to simulate transactions that would cause the accounts data size to decrease. IMO this seems sufficient for this test but wanted to run it by you to make sure. Wdyt?

Reply:
Done! I've implemented this simpler option.

Reply:
Redone. Implemented a new mock realloc instruction to perform a resize that shrinks the account.